File: alien_numbers/alien_numbers.py
def source_to_decimal(alien_number, source_language):
    N_base = len(source_language)
    result = 0
    for an in alien_number:
        result = N_base*result + source_language.find(an)
    return result

def decimal_to_target(decimal, target_language):
    N_base = len(target_language)
    translation = ""
    while decimal != 0:
        remainder = decimal % N_base
        decimal //= N_base
        translation = target_language[remainder] + translation
    return translation

N = int(input())
for i in range(N):
    alien_number, source_language, target_language = input().split()
    decimal_repr = source_to_decimal(alien_number, source_language)
    translation = decimal_to_target(decimal_repr, target_language)
    print(f"Case #{i + 1}: {translation}")

File: alien_numbers/alien_numbers.rb
def source_to_decimal(alien_number, source_language)
    n_base = source_language.length
    result = 0
    for an in alien_number.each_char do
        result = n_base*result + source_language.index(an)
    end
    return result
end

def decimal_to_target(decimal, target_language)
    n_base = target_language.length
    translation = ""
    while decimal != 0 do
        remainder = decimal % n_base
        decimal /= n_base
        translation = target_language[remainder] + translation
    end
    return translation
end

n = gets.to_i
for i in 1..n
    alien_number, source_language, target_language = gets.split()
    decimal_repr = source_to_decimal(alien_number, source_language)
    puts decimal_repr
    translation = decimal_to_target(decimal_repr, target_language)
    print("Case \##{i}: #{translation}\n")
end
